# Patients No_Show Appointment Dataset
#### by Nwangene Sobe-Olisa Andrew

### Table of Content
* [Introduction](#introduction)
* [Data Wrangling](#data-wrangling)
* [Exploratory Data Analysis](#exploratory-data-analysis)
* [Predictive Analysis](../script/show_prediction.ipynb)
* [Presentation](#conclusions)

## Introduction
Following the description in the accompanying README file, this project is aimed at analyzing medical appointment data from state sources in Brazil. The project is divided into subcatgorizes as seen in the table of content. In this introduction section, i'll introduce the dataset, assessing it's characteristics and contents.
Firstly, let's import the neccessary packages for this project:

In [1]:
#importing packages for this project and set plots to be embedded inline
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as snb 
%matplotlib inline 
import datetime as dt
import warnings
warnings.filterwarnings('ignore')

#loading the dataset into pandas dataframe
appointment = pd.read_csv('../data/noshowappointments-kagglev2-may-2016.csv')
appointment.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   PatientId       110527 non-null  float64
 1   AppointmentID   110527 non-null  int64  
 2   Gender          110527 non-null  object 
 3   ScheduledDay    110527 non-null  object 
 4   AppointmentDay  110527 non-null  object 
 5   Age             110527 non-null  int64  
 6   Neighbourhood   110527 non-null  object 
 7   Scholarship     110527 non-null  int64  
 8   Hipertension    110527 non-null  int64  
 9   Diabetes        110527 non-null  int64  
 10  Alcoholism      110527 non-null  int64  
 11  Handcap         110527 non-null  int64  
 12  SMS_received    110527 non-null  int64  
 13  No-show         110527 non-null  object 
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


A look at the output shows that the dataset has 14 columns and 110,527 rows. Column description is as follows:
### Dataset Description 

This dataset is a record of medical appointment in Brazil. it contains 100,000 medical appointment data and i'll be looking to determine a partner based on the dataset. A description of the dataset is contained in the column names which can be explained as follows:

* `Patient_ID` is the unique Identifier for each patient.
* `Appointment_ID` is the unique appointment identifier.
* `Gender` records the gender of the applicant
* `Scheduled_Day` tells us on what day the patient set up their appointment.
* `appointment_Day` tells us the appointment date.
* `Neighborhood` indicates the location of the hospital.
* `Scholarship` indicates whether or not the patient is enrolled in Brasilian welfare program.
* `Hipertension` indicates if the applicant is hypertensive or not.
* `Diabetes` indicates the applicants diabetes status.
* `Alcoholism` indicates the applicants alcohol use.
* `Handcap` indicates the applicants physical attributes.
* `SMS_received` indicates if the applicant received a message reminder for the appointment.
* `No_show` records the outcomes of the appointment.
> The goal of this analysis is to observe variables and how they affect if an individual shows up for an appointment or not. Along the way, we'll look at the independent variables, individually or collectively, to ascertain their importance to the dependent variable(No Show). The relationships will be presented in a slide and also used in a predictive analysis. 

Next, having imported the neccessary packages and the dataset, let's assess and clean up the data in the [Data Wrangling](#data-wrangling) section.


# Data Wrangling
### Data Assessing
Let's assess the dataset, checking for data quality and inconsistency issues.

In [2]:
appointment.sample(n=4)#.head()

PatientId  AppointmentID Gender          ScheduledDay  \
37463  6.242646e+10        5656949      M  2016-05-04T08:46:06Z   
22424  1.756976e+13        5753401      M  2016-05-31T14:01:53Z   
89150  6.169878e+13        5755359      M  2016-06-01T07:18:40Z   
84497  3.689496e+13        5731570      F  2016-05-24T10:13:49Z   

             AppointmentDay  Age      Neighbourhood  Scholarship  \
37463  2016-05-04T00:00:00Z   17  JESUS DE NAZARETH            0   
22424  2016-05-31T00:00:00Z   41     NOVA PALESTINA            0   
89150  2016-06-01T00:00:00Z   15          CARATOÍRA            0   
84497  2016-06-02T00:00:00Z   19     NOVA PALESTINA            0   

       Hipertension  Diabetes  Alcoholism  Handcap  SMS_received No-show  
37463             0         0           0        0             0      No  
22424             1         0           0        0             0      No  
89150             1         0           0        0             0      No  
84497             0         0           0        0             1      No

In [3]:
print(f'Dataset size: {appointment.shape}')# returns the number of rows and columns in dataframe
print('#' *70) # to separate the different outputs
print(f' Number of unique Values: {appointment.nunique()}')# returns number of unique values in each column
print('#' *70) # to separate the different outputs.
print(f' Number of null values in each column: {appointment.isnull().sum()}')# returns sum of null values in each column

Dataset size: (110527, 14)
######################################################################
 Number of unique Values: PatientId          62299
AppointmentID     110527
Gender                 2
ScheduledDay      103549
AppointmentDay        27
Age                  104
Neighbourhood         81
Scholarship            2
Hipertension           2
Diabetes               2
Alcoholism             2
Handcap                5
SMS_received           2
No-show                2
dtype: int64
######################################################################
 Number of null values in each column: PatientId         0
AppointmentID     0
Gender            0
ScheduledDay      0
AppointmentDay    0
Age               0
Neighbourhood     0
Scholarship       0
Hipertension      0
Diabetes          0
Alcoholism        0
Handcap           0
SMS_received      0
No-show           0
dtype: int64


In [4]:
# checking for duplicates in columns
appointment.duplicated().sum()

0

In [5]:
appointment.describe()

PatientId  AppointmentID            Age    Scholarship  \
count  1.105270e+05   1.105270e+05  110527.000000  110527.000000   
mean   1.474963e+14   5.675305e+06      37.088874       0.098266   
std    2.560949e+14   7.129575e+04      23.110205       0.297675   
min    3.921784e+04   5.030230e+06      -1.000000       0.000000   
25%    4.172614e+12   5.640286e+06      18.000000       0.000000   
50%    3.173184e+13   5.680573e+06      37.000000       0.000000   
75%    9.439172e+13   5.725524e+06      55.000000       0.000000   
max    9.999816e+14   5.790484e+06     115.000000       1.000000   

        Hipertension       Diabetes     Alcoholism        Handcap  \
count  110527.000000  110527.000000  110527.000000  110527.000000   
mean        0.197246       0.071865       0.030400       0.022248   
std         0.397921       0.258265       0.171686       0.161543   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       4.000000   

        SMS_received  
count  110527.000000  
mean        0.321026  
std         0.466873  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         1.000000  
max         1.000000

In [6]:
appointment.describe(include=['O'])

Gender          ScheduledDay        AppointmentDay   Neighbourhood  \
count   110527                110527                110527          110527   
unique       2                103549                    27              81   
top          F  2016-05-06T07:09:54Z  2016-06-06T00:00:00Z  JARDIM CAMBURI   
freq     71840                    24                  4692            7717   

       No-show  
count   110527  
unique       2  
top         No  
freq     88208

In [7]:
appointment['Age'].nsmallest(10)

99832   -1
59       0
63       0
64       0
65       0
67       0
89       0
101      0
104      0
132      0
Name: Age, dtype: int64

These cell blocks above have shown the different data quality issues plaquing the datasets. A large chunck of the dataset have categorical values and as such should be changed to categorical datatype. The dataset seems to have no Null values but the minimum age shows '-1', which is an obvious erroneous data. This correction, data type unconformity, column names not in line with PEP 8 standards and finally irrelevant columns (patient_Id and Appiontment_id) will be fixed next.

### Data Cleaning
First, we create a copy of the dataset. working with a copy of the dataset is best practice as allows the original dataset retain it's original structure.
Here, i'll:

1) Drop columns that are irrelevant to our analysis (ID columns)

2) data type conformity

3) check for duplicated data

4) Rename the column name in line with PEP 8


From the output table above, it's observed that 'Age' has a minimum value of '-1'. this cant be true. and most be corrected. 
This correction, data type unconformity, column names not in line with PEP 8 standards and finally irrelevant columns (patient_Id and Appiontment_id) will be fixed in the data cleaning subsection.

### Data Cleaning: 
First, we create a copy of the dataset. working with a copy of the dataset is best practice as allows the original dataset retain it's original structure


In [8]:
df = appointment.copy()

Data cleaning processes to be applied here includes:

1) Changing column names to match PEP 8 standards.

2) Changing 'Scheduled_day' and 'Appointment_day' datatypes to match their content.

3) Removal of irrelevant datasets like 'Patient_Id' and 'Appointment_Id'.

4) Then fixing the issue of '-1' as the min age in the 'Age' column.

5) Let's change the 'No_show' column content from 'yes','no' to '1','0'. This will make for easy computation. 

In [9]:
# changing column names
# i'll be renaming the 'No_show' column to Show. 
labels =['Patient_ID', 'Appointment_ID', 'Gender', 'Scheduled_day', 'Appointment_day', 'Age', 'Neighbourhood', 
'Scholarship', 'Hipertension', 'Diabetes', 'Alcoholism', 'Handcap', 'SMS_received', 'Show']
df = pd.read_csv('../data/noshowappointments-kagglev2-may-2016.csv', header=0, names=labels)
df.head()

Patient_ID  Appointment_ID Gender         Scheduled_day  \
0  2.987250e+13         5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14         5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12         5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11         5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12         5642494      F  2016-04-29T16:07:23Z   

        Appointment_day  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received Show  
0         0           0        0             0   No  
1         0           0        0             0   No  
2         0           0        0             0   No  
3         0           0        0             0   No  
4         1           0        0             0   No

In [10]:
#changing Scheduled_day and Appointment_day datatype to timestamps
df['Scheduled_day'] = pd.to_datetime(df['Scheduled_day'])
df['Appointment_day'] = pd.to_datetime(df['Appointment_day'])

# removing the time in the datetime column ('Scheduled_day') to match the 'Appointment_day' column
df['Scheduled_day'] = pd.to_datetime(df['Scheduled_day'].dt.date)
df['Appointment_day'] = pd.to_datetime(df['Appointment_day'].dt.date)

# Extracting the 'Year','Month' and 'day' from Date Column.
df['scheduled_month'] = df['Scheduled_day'].dt.month_name()
df['scheduled_day'] = df['Scheduled_day'].dt.day_name()

df['appointment_month'] = df['Appointment_day'].dt.month_name()
df['appointment_day'] = df['Appointment_day'].dt.day_name()
print(df.dtypes)

df.head()

Patient_ID                  float64
Appointment_ID                int64
Gender                       object
Scheduled_day        datetime64[ns]
Appointment_day      datetime64[ns]
Age                           int64
Neighbourhood                object
Scholarship                   int64
Hipertension                  int64
Diabetes                      int64
Alcoholism                    int64
Handcap                       int64
SMS_received                  int64
Show                         object
scheduled_month              object
scheduled_day                object
appointment_month            object
appointment_day              object
dtype: object


Patient_ID  Appointment_ID Gender Scheduled_day Appointment_day  Age  \
0  2.987250e+13         5642903      F    2016-04-29      2016-04-29   62   
1  5.589978e+14         5642503      M    2016-04-29      2016-04-29   56   
2  4.262962e+12         5642549      F    2016-04-29      2016-04-29   62   
3  8.679512e+11         5642828      F    2016-04-29      2016-04-29    8   
4  8.841186e+12         5642494      F    2016-04-29      2016-04-29   56   

       Neighbourhood  Scholarship  Hipertension  Diabetes  Alcoholism  \
0    JARDIM DA PENHA            0             1         0           0   
1    JARDIM DA PENHA            0             0         0           0   
2      MATA DA PRAIA            0             0         0           0   
3  PONTAL DE CAMBURI            0             0         0           0   
4    JARDIM DA PENHA            0             1         1           0   

   Handcap  SMS_received Show scheduled_month scheduled_day appointment_month  \
0        0             0   No           April        Friday             April   
1        0             0   No           April        Friday             April   
2        0             0   No           April        Friday             April   
3        0             0   No           April        Friday             April   
4        0             0   No           April        Friday             April   

  appointment_day  
0          Friday  
1          Friday  
2          Friday  
3          Friday  
4          Friday

In [11]:
# replacing the month and day columns with interger values as this is an ML project
df.scheduled_month.replace(to_replace=['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 
'September', 'October', 'November', 'December'], value=[1,2,3,4,5,6,7,8,9,10,11,12], inplace=True)
df.scheduled_day.replace(to_replace=['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday','Friday', 'Saturday'],
value=['1', '2', '3', '4', '5', '6', '7'], inplace=True)
df.appointment_month.replace(to_replace=['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 
'September', 'October', 'November', 'December'], value=[1,2,3,4,5,6,7,8,9,10,11,12], inplace=True)
df.appointment_day.replace(to_replace=['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday','Friday', 'Saturday'],
value=['1', '2', '3', '4', '5', '6', '7'], inplace=True)

df.head()

Patient_ID  Appointment_ID Gender Scheduled_day Appointment_day  Age  \
0  2.987250e+13         5642903      F    2016-04-29      2016-04-29   62   
1  5.589978e+14         5642503      M    2016-04-29      2016-04-29   56   
2  4.262962e+12         5642549      F    2016-04-29      2016-04-29   62   
3  8.679512e+11         5642828      F    2016-04-29      2016-04-29    8   
4  8.841186e+12         5642494      F    2016-04-29      2016-04-29   56   

       Neighbourhood  Scholarship  Hipertension  Diabetes  Alcoholism  \
0    JARDIM DA PENHA            0             1         0           0   
1    JARDIM DA PENHA            0             0         0           0   
2      MATA DA PRAIA            0             0         0           0   
3  PONTAL DE CAMBURI            0             0         0           0   
4    JARDIM DA PENHA            0             1         1           0   

   Handcap  SMS_received Show  scheduled_month scheduled_day  \
0        0             0   No                4             6   
1        0             0   No                4             6   
2        0             0   No                4             6   
3        0             0   No                4             6   
4        0             0   No                4             6   

   appointment_month appointment_day  
0                  4               6  
1                  4               6  
2                  4               6  
3                  4               6  
4                  4               6

In [12]:
# changing strings value ('yes','no') in 'No_show to int values ('0','1')
# with this, every '1' value in the 'show' dataset indicates that the patient showed up
# and '0' means the patient didnt.

df['Show'].replace(to_replace=['Yes', 'No'], value=[0,1], inplace=True)
df.head()

Patient_ID  Appointment_ID Gender Scheduled_day Appointment_day  Age  \
0  2.987250e+13         5642903      F    2016-04-29      2016-04-29   62   
1  5.589978e+14         5642503      M    2016-04-29      2016-04-29   56   
2  4.262962e+12         5642549      F    2016-04-29      2016-04-29   62   
3  8.679512e+11         5642828      F    2016-04-29      2016-04-29    8   
4  8.841186e+12         5642494      F    2016-04-29      2016-04-29   56   

       Neighbourhood  Scholarship  Hipertension  Diabetes  Alcoholism  \
0    JARDIM DA PENHA            0             1         0           0   
1    JARDIM DA PENHA            0             0         0           0   
2      MATA DA PRAIA            0             0         0           0   
3  PONTAL DE CAMBURI            0             0         0           0   
4    JARDIM DA PENHA            0             1         1           0   

   Handcap  SMS_received  Show  scheduled_month scheduled_day  \
0        0             0     1                4             6   
1        0             0     1                4             6   
2        0             0     1                4             6   
3        0             0     1                4             6   
4        0             0     1                4             6   

   appointment_month appointment_day  
0                  4               6  
1                  4               6  
2                  4               6  
3                  4               6  
4                  4               6

In [13]:
df.Neighbourhood.nunique()

81

Finally, to fix the min Age error. I'll have to identify the row with '-1' as it's value and drop it from the dataset. To identify this row, i'll have to query the 'age' dataset for that value.

In [ ]:
# dropping the erroneous Age data
df = df[df['Age'] >= 0]

In [ ]:
# to remove the irrelevant columns
df.drop(['Appointment_day', 'Scheduled_day', 'Patient_ID', 'Appointment_ID'], axis = 1, inplace = True)
df.head()

## Feature Engineering
Having dealt with the data quality and tirdiness issues, a bit of feature engineering will be needed since creating a machine learning model is part of the analysis objectives. For this, i'll turn the neigborhood variable into numbers and then group the neigborhood into 8. This will come in handy in model building and training.

In [ ]:
# working on the neigbhorhood column
neigbhorhood=pd.get_dummies(df.Neighbourhood, drop_first=True)
neigbhorhood

In [ ]:
# grouping the reason neigbhorhood data
neigbhorhood_1 = neigbhorhood.iloc[:, :11].max(axis=1)
neigbhorhood_2 = neigbhorhood.iloc[:, 11:21].max(axis=1)
neigbhorhood_3 = neigbhorhood.iloc[:, 21:31].max(axis=1)
neigbhorhood_4 = neigbhorhood.iloc[:, 31:41].max(axis=1)
neigbhorhood_5 = neigbhorhood.iloc[:, 41:51].max(axis=1)
neigbhorhood_6 = neigbhorhood.iloc[:, 51:61].max(axis=1)
neigbhorhood_7 = neigbhorhood.iloc[:, 61:71].max(axis=1)
neigbhorhood_8 = neigbhorhood.iloc[:, 71:].max(axis=1)

In [ ]:
df = pd.concat([df, neigbhorhood_1, neigbhorhood_2, neigbhorhood_3, neigbhorhood_4, neigbhorhood_5, neigbhorhood_6, neigbhorhood_7, neigbhorhood_8], axis=1)
df

Next, we categorize the age variable according to generations. This makes for easy computing.

In [ ]:
appointment.Age.max()

In [ ]:
#using pandas_cut function 
#create the bin_edges that will be used to cut the data into groups.
bin_edges = [-1.0, 22, 38.0, 54.0, 115.0]

#create labels for the new categories.
# 1 (Gen_Z+), 2 (Milennials), 3 (Gen_X), 4 (Bloomers+)
bin_names = ['1', '2', '3', '4']

# puting the pandas_cut function to use
df['age_groups'] = pd.cut(df['Age'], bin_edges, labels=bin_names)

#drop age column
# # checks for the successful creation of 'Age_group' column
df.head()


In [ ]:
df.columns.values

In [ ]:
# renaming the columns
column_names = ['gender', 'age', 'neighbourhood', 'scholarship', 'hypertension',
       'diabetes', 'alcoholism', 'handicap', 'sms_received', 'show',
       'scheduled_month', 'scheduled_day', 'appointment_month',
       'appointment_day', 'neigbhorhood_1', 'neigbhorhood_2', 'neigbhorhood_3', 
       'neigbhorhood_4', 'neigbhorhood_5', 'neigbhorhood_6', 'neigbhorhood_7', 'neigbhorhood_8', 'age_groups']

df.columns = column_names
df.head()

In [ ]:
df.drop(['neighbourhood', 'age'], axis = 1, inplace = True)

In [ ]:
df.head()

In [ ]:
df.dtypes

In [ ]:
df[['age_groups', 'scholarship', 'hypertension', 'diabetes', 'alcoholism', 'handicap', 
'sms_received', 'show', 'scheduled_month', 'appointment_month', 'scheduled_day', 
'appointment_day']].corr().style.background_gradient(cmap='coolwarm')

This correlation chart shows some correlation between the variables. Scheduled month appears to have the best(though weak) correlation to the dependent variable amongst the indpendent variables. Also, Age, hypertension and diabetes has a good correlation amongst each other.

## Exploratory Data Analysis
Here, I'll be using python powerful tools like: pandas, numpy, matplotlib to run analysis, find patterns and visualize relationships with the goal of addressing the research questions. Recall that this dataset can be classed into two(2), the dependent variable (Show) and the independent variables (Dataset - show). Let's start with the dependent variable:

1) What is the spread of the dependent variable. Knowing this will be of great importance when building a model off the dataset. A suitable dataset for model development should have a reasonable amount of both outcomes of the dependent variable, that way, the model can perform well in both scenario.

In [ ]:
ax = snb.countplot(data = df, x = 'show',  color= snb.color_palette()[3], order = df.show.value_counts().index)

total = len(df)
for p in ax.patches:
    percentage = '{:.3f}%'.format(100 * p.get_height()/total)
    x = p.get_x() + p.get_width()/2
    y = p.get_height()+0.5
    ax.annotate(percentage, (x, y),ha='center')
plt.xlabel('show')
plt.title('Spread of the dependent variable')
plt.show()

Fair enough, 79% of the patients showed up for an appointment while 20% didn't.

### 1) Are patients more likely to turn up for an appointment if its free?

The dataframe shows that some patients got funding whilst other didn't. We are going to see how 'scholarship' affects the turn up (Show') level of patients. To achieve this, first we group 'Scholarship' dataset and 'No_show' dataset. Then calculate for the mean values.  

In [ ]:
def group(df, col_1, col_2):
    df_c = df.groupby(col_1)[col_2].mean()
    return df_c

# grouping 'scholarship' and 'Show' and finding the sum of 'show' based on 'Scholarship'
group(df, 'scholarship', 'show')

This pivot table shows that government scholarship on it's own doesn't result to more patients attending appointment, as 80% of patients with no scholarship showed up while only 76% of those with scholarship did. This means that patients on scholarship that didn't attend (24%) is more than patients without scholarship that didn't attend (<20%).

In [ ]:
# to reduce code repetitions, a function will be created 
# to be called on for bivariant plots.
df_s = df.groupby('scholarship')['show'].value_counts()
def plot(df, xlabel, ylabel, title):
    df.plot(kind='bar', title=title)
    plt.xlabel("{}".format(xlabel), fontsize=18)
    plt.ylabel("{}".format(ylabel), fontsize=18)
    plt.legend;
    plt.show()

plot(df_s, 'scholarship', 'show', 'Relationship Scholarship and Attendance')

where:
* (0,1) = No scholarship Attendees
* (0,0) = No scholarhip None Attendees
* (1,0) = Scholarship None attendees
* (1,1) = Scholarship Attendees.

### 2) Since money isn't an important factor, could it be prevalent medical history ?
Let's see how patients with any medical conditions reacted to their medical appointment.

In [ ]:
print(group(df, 'hypertension', 'show'))
print('#' * 50)
print(group(df, 'diabetes', 'show'))
print('#' * 50)
print(group(df, 'handicap', 'show'))
print('#' * 50)
print(group(df, 'alcoholism', 'show'))

The pivot tables above shows that patients with any of the illness above has a higher chance of showing up for an appointment than their counterparts without any of the aliment. So, it shows that having any of the above health challenge is a factor in this analysis. Next, we'll consider the residential area of the patients as contained in `neighborhood` columns. 

In [ ]:
print(group(df, 'neigbhorhood_1', 'show'))
print('#' * 50)
print(group(df, 'neigbhorhood_2', 'show'))
print('#' * 50)
print(group(df, 'neigbhorhood_3', 'show'))
print('#' * 50)
print(group(df, 'neigbhorhood_4', 'show'))

In [ ]:
print(group(df, 'neigbhorhood_5', 'show'))
print('#' * 50)
print(group(df, 'neigbhorhood_6', 'show'))
print('#' * 50)
print(group(df, 'neigbhorhood_7', 'show'))
print('#' * 50)
print(group(df, 'neigbhorhood_8', 'show'))

These two cells shows that patients from some places (neigbhoods 2,5,6,7,8) have a higher chance of showing up for an appointment than others, with neigbhorhood 2 having the highest percentage(81%). So now we have some independent variables that have an effect on the possibility of a show up (aliments and Neigbhorhood 2,5,6,7,8), lets consider other variables too.

In [ ]:
df.columns.values

In [ ]:
print(group(df, 'gender', 'show'))
print('#' * 50)
print(group(df, 'age_groups', 'show'))
print('#' * 50)
print(group(df, 'scheduled_month', 'show'))
print('#' * 50)
print(group(df, 'sms_received', 'show'))

* Interestly, this pivot tables shows that male are more likely to show up then females and as patients age, the higher the chance of them showing up for an appointment(expect for milennials whom normally assume to be healthiest).
* The scheduled month also brings an interesting information, appointment scheduled on the 5,6 and 11th month show great chance of a show up(>80%), while 3rd, 4th and 12th month are the least. interestly, every event scheduled by the 11th month of the year experienced a 100% turn up because the data only has a single event on that month as seen in the query below. also note that no data for months 7-10.
* Apparently, notifying patients of an appointment doesn't improve the chance of the patient showing up for the appointment. Actually patients who didn't receive and sms showed up more than those who did( maybe it has to do with when the text was sent).

In [ ]:
#query for data scheduled on the 11th month of the year
df.query('scheduled_month == "11"')

In [ ]:
# pivot tables for appointment_day, scheduled_day, appointment_month and how they relate to showing up 
print(group(df, 'appointment_day', 'show'))
print('#' * 50)
print(group(df, 'scheduled_day', 'show'))
print('#' * 50)
print(group(df, 'appointment_month', 'show'))

* Here, we can notice that the first day of the week(Sundays) aren't recorded, most likely cause brazil is a religious country, so no appointments and scheduling calls on that day. 
* Furthermore,it's observed that appointments for the 4th and 5th days of the week sees a higher show rate(>80%), while scheduling days of 2nd and 7th (representing Monday and Saturday) recorded the highest rate of showing up with saturday having the highest 95%. Querrying the scheduled_day for 'saturday' shows about 24 patients under this class. Why is saturday having such a high turn out? more data is needed.
* Also, on the appointment month variable, it shows that 4th and 6th month of the year has an increasing percentage of showing up.

In [ ]:
# querrying scheduledday column for saturday.
df.query('scheduled_day == "7"')

In [ ]:
df.columns

So, we now know variables that effect the ability of patients to show up, namely:
* Diabetes
* Hypertension
* Handicap
* Neigbhorhood_2
* Neigbhorhood_5
* Neigbhorhood_7
* Neigbhorhood_8

And we've also seen some insights into how the dependent column(show) reacts across the different categories in:
* scheduled_month
* appointment_month
* scheduled_day
* appointment_day
* age_groups 

In [ ]:
# pivpt table for 'age group', 'diabetes' against 'show'
da_pivot = pd.crosstab(index=df["age_groups"], columns=df["diabetes"], values=df["show"], aggfunc=np.mean)
da_pivot

Interestly, diabetics across all age groups are less likely to show up than non-diabetics(excepts for a slight increase in milennials). Recall that we earlier established that diabetics are more likely to show than non-diabetics, but it appears this uptick by diabetics has nothing to do with age. Let's compare across other factors.

In [ ]:
# pivpt table of handicap and diabetes against show
hd_pivot = pd.crosstab(index=df["handicap"], columns=df["diabetes"], values=df["show"], aggfunc=np.mean)
hd_pivot

While the dataset doesn't contain diabetes with handicaps 3 & 4, othe diabetes with other forms of handicaps 1&2 are more likely to show up than their non-diabetic counterparts.

In [ ]:
# pivot table of hypertension and diabetes against show
hd_pivot = pd.crosstab(index=df["hypertension"], columns=df["diabetes"], values=df["show"], aggfunc=np.mean)
hd_pivot

Hypertensive Non-diabetics are more likely to show up than their diabetic counterpart. 

In [ ]:
# pivot table between neigbhorhood_2 and diabetes against show
nd_pivot = pd.crosstab(index=df["neigbhorhood_2"], columns=df["diabetes"], values=df["show"], aggfunc=np.mean)
nd_pivot

In [ ]:
# pivot table between neigbhorhood_5 and diabetes against show
pivot_nd_5 = pd.crosstab(index=df["neigbhorhood_5"], columns=df["diabetes"], values=df["show"], aggfunc=np.mean)
pivot_nd_5

In [ ]:
# querry for diabetics within neighborhood 5
df.loc[(df['neigbhorhood_5'] == 1) &(df['diabetes'] == 1), :]

There seem to be a large jump in show up level for diabetes living within neighborhood 5 and this is more interest giving that over 500 patients fall under this class, so its not a case of limited data.

In [ ]:
# # pivot table between neigbhorhood_7 and diabetes against show
pivot_nd_7 = pd.crosstab(index=df["neigbhorhood_7"], columns=df["diabetes"], values=df["show"], aggfunc=np.mean)
pivot_nd_7

In [ ]:
# pivot table between neigbhorhood_8 and diabetes against show
pivot_nd_8 = pd.crosstab(index=df["neigbhorhood_8"], columns=df["diabetes"], values=df["show"], aggfunc=np.mean)
pivot_nd_8

In [ ]:
pivot_nd_8 = pd.crosstab(index=df["diabetes"], columns=df["appointment_month"], values=df["show"], aggfunc=np.mean)
pivot_nd_8

Diabetics are more likely to show up for an appointment than non-diabetics across all appointment month.

In [ ]:
pivot_ds = pd.crosstab(index=df["diabetes"], columns=df["scheduled_month"], values=df["show"], aggfunc=np.mean)
pivot_ds

Diabetes within months 2 and 3 are less likely to show up than non-diabetisc. Also recall that months 2&3 have a lower show up average.

In [ ]:
df.loc[(df['scheduled_month'] == 2) &(df['diabetes'] == 1), :]

In [ ]:
pivot_ds = pd.crosstab(index=df["diabetes"], columns=df["scheduled_day"], values=df["show"], aggfunc=np.mean)
pivot_ds

%

In [ ]:
pivot_ds = pd.crosstab(index=df["diabetes"], columns=df["appointment_day"], values=df["show"], aggfunc=np.mean)
pivot_ds

In [ ]:
print(pd.crosstab(index=df["hypertension"], columns=df["age_groups"], values=df["show"], aggfunc=np.mean))


In [ ]:
print(pd.crosstab(index=df["hypertension"], columns=df["handicap"], values=df["show"], aggfunc=np.mean))


In [ ]:
print(pd.crosstab(index=df["hypertension"], columns=df["neigbhorhood_2"], values=df["show"], aggfunc=np.mean))


In [ ]:
print(pd.crosstab(index=df["hypertension"], columns=df["neigbhorhood_5"], values=df["show"], aggfunc=np.mean))


In [ ]:
print(pd.crosstab(index=df["hypertension"], columns=df["neigbhorhood_7"], values=df["show"], aggfunc=np.mean))
print(pd.crosstab(index=df["hypertension"], columns=df["neigbhorhood_8"], values=df["show"], aggfunc=np.mean))


In [ ]:
pivot = pd.crosstab(index=df["hypertension"], columns=df["scheduled_month"], values=df["show"], aggfunc=np.mean)
pivot

In [ ]:
print(pd.crosstab(index=df["hypertension"], columns=df["appointment_month"], values=df["show"], aggfunc=np.mean))
print(pd.crosstab(index=df["hypertension"], columns=df["scheduled_day"], values=df["show"], aggfunc=np.mean))


In [ ]:
print(pd.crosstab(index=df["hypertension"], columns=df["appointment_day"], values=df["show"], aggfunc=np.mean))


In [ ]:
check1= pd.crosstab(df.gender, df.show).plot(kind='bar')

In [ ]:
check = pd.crosstab(df.gender, df.show)
check

In [ ]:
df.drop(['scholarship', 'alcoholism', 'sms_received', 'neigbhorhood_1', 
'neigbhorhood_3', 'neigbhorhood_3', 'neigbhorhood_4', 'neigbhorhood_6'], axis=1, inplace=True)

df.to_csv('../data/enhanced.csv', index=False)

## Conclusions

From the analysis done through the questions answered, it shows the following:

a) Gender is a factor in determining show up; females are more likely than males

b) Aged patients are most likely to show up than other age groups.

C) Though Diabetic patients are more likely to show up than Non-diabetic patients, Hypertensive patients are more likely to show up than the patients with diabetes.

d) Scholarship and SMS isn't a factor in determining patients show up.

### Limitations
1) Dropping values due to errors can always skew analysis. Though the number of dropped rows here are minute in comparison to the total data used.

2) I should look to improving my model for better accuracy over the 89% recorded.

3) The data set summaries health conditions in "1" and "0" format, giving no  consideration for different severity of health conditions and this can go a long way in justifying committment to appointment.


In [ ]:
# from subprocess import call
# call(['python', '-m', 'nbconvert', 'Investigate_a_Dataset.ipynb'])

[back to top](#introduction)